File: tally-android/app/src/androidTest/java/com/tally/app/OfflineTests.kt
package com.tally.app

import androidx.compose.ui.test.junit4.createAndroidComposeRule
import androidx.compose.ui.test.onNodeWithTag
import androidx.compose.ui.test.onNodeWithText
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.tally.app.pages.AuthPage
import com.tally.app.pages.ChallengeDialogPage
import com.tally.app.pages.DashboardPage
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

/**
 * Tests for offline user experience.
 * Maps to cucumber/02-offline-user-experience.feature
 */
@RunWith(AndroidJUnit4::class)
class OfflineTests {
    
    @get:Rule
    val composeRule = createAndroidComposeRule<MainActivity>()
    
    private val authPage by lazy { AuthPage(composeRule) }
    private val dashboardPage by lazy { DashboardPage(composeRule) }
    private val challengeDialog by lazy { ChallengeDialogPage(composeRule) }
    
    /**
     * Navigate to dashboard, either by entering local-only mode from sign-in
     * or continuing if already on dashboard.
     */
    private fun navigateToDashboard() {
        composeRule.waitForIdle()
        
        // Try to find sign-in screen first
        val isOnSignIn = try {
            composeRule.onNodeWithTag("sign_in_screen").assertExists()
            true
        } catch (e: AssertionError) {
            false
        }
        
        if (isOnSignIn) {
            authPage.tapContinueWithoutAccount()
            composeRule.waitForIdle()
        }
        
        // Verify we're on dashboard
        composeRule.onNodeWithTag("dashboard").assertExists()
    }
    
    // MARK: - Starting as Offline User
    
    @Test
    fun testLaunchingAppWithoutAccount() {
        navigateToDashboard()
        
        // Should be on dashboard with empty state or create button
        val hasDashboard = try {
            composeRule.onNodeWithTag("dashboard").assertExists()
            true
        } catch (e: AssertionError) {
            false
        }
        
        assert(hasDashboard) { "Should be on dashboard" }
    }
    
    @Test
    fun testCanCreateChallengeWhileOffline() {
        navigateToDashboard()
        
        // Create a challenge (should work in local-only mode)
        dashboardPage.tapCreateChallenge()
        composeRule.waitForIdle()
        
        challengeDialog.fillChallenge(name = "Offline Challenge", target = "1000")
        challengeDialog.tapSave()
        
        // Should be visible immediately
        composeRule.waitForIdle()
        dashboardPage.assertChallengeExists("Offline Challenge")
    }
    
    // MARK: - Sync Status
    
    @Test
    fun testViewSyncStatusAsOfflineUser() {
        navigateToDashboard()
        
        // Look for sync status indicator showing LOCAL_ONLY
        val syncStatus = composeRule.onNodeWithTag("sync_status")
        syncStatus.assertExists()
        
        // Should show "Local only" text
        composeRule.onNodeWithText("Local only", substring = true).assertExists()
    }
}

File: ios/App/UITests/OfflineTests.swift
import XCTest

/// Tests for offline user experience
/// Maps to cucumber/02-offline-user-experience.feature
final class OfflineTests: TallyUITestCase {
    
    lazy var dashboardPage = DashboardPage(app: app)
    lazy var challengeDialog = ChallengeDialogPage(app: app)
    
    // MARK: - Starting as Offline User
    
    func testLaunchingAppWithoutAccount() throws {
        // App should launch and show dashboard
        XCTAssertTrue(
            dashboardPage.createChallengeButton.waitForExistence(timeout: 10),
            "Should be able to use app without account"
        )
    }
    
    func testCanCreateChallengeWhileOffline() throws {
        _ = waitForElement(dashboardPage.createChallengeButton)
        
        // Create a challenge (should work offline)
        dashboardPage.tapCreateChallenge()
        // Only fill name - use default target to avoid input issues
        challengeDialog.fillChallenge(name: "Offline Challenge")
        challengeDialog.tapSaveAndWaitForDismiss()
        
        // Should be visible after UI updates
        dashboardPage.assertChallengeExists(named: "Offline Challenge")
    }
    
    func testDataPersistsAcrossAppRestart() throws {
        // Create a challenge
        _ = waitForElement(dashboardPage.createChallengeButton)
        dashboardPage.tapCreateChallenge()
        challengeDialog.fillChallenge(name: "Persistence Test", target: "500")
        challengeDialog.tapSaveAndWaitForDismiss()
        dashboardPage.assertChallengeExists(named: "Persistence Test")
        
        // Terminate and relaunch
        app.terminate()
        app.launch()
        
        // Challenge should still exist
        _ = waitForElement(dashboardPage.createChallengeButton, timeout: 10)
        dashboardPage.assertChallengeExists(named: "Persistence Test")
    }
    
    // MARK: - Sync Status
    
    func testViewSyncStatusAsOfflineUser() throws {
        _ = waitForElement(dashboardPage.createChallengeButton)
        
        // Look for sync status indicator
        if dashboardPage.syncStatus.waitForExistence(timeout: 5) {
            let statusText = dashboardPage.syncStatus.label
            // Should indicate local/offline mode
            XCTAssertTrue(
                statusText.lowercased().contains("local") || 
                statusText.lowercased().contains("offline"),
                "Sync status should indicate local/offline mode"
            )
        }
    }
}
